### ISSUE TITLE
avoid_returning_this rule should not apply to operator overloading

### ISSUE THREAD

**Issue by aemino** [dart-lang/linter#791]
I have a class in my code called `Route`, and one of the operators looks like this:
```dart
Route operator +(String name) {
  path.add(name);
  return this;
}
```

Operators frequently return the original object being operated on with a modification. I don't think the `avoid_returning_this` rule should apply to operator overloads.

**Comment by bwilkerson** [comment]
I think a fairly compelling argument can be made that most developers expect that operators should not have side effects, and as a result they rarely do. Hence I disagree with your claim that "operators _frequently_ return the original object being operated on with a modification." (italics added).

**Comment by aemino** [comment]
I probably made too broad of a claim, but I still don't believe overload operators should be included with the rule.  As far as I am aware, the rule description seems to indicate its purpose is to discourage returning this just to chain operations (as that's what the cascade operator is for.)  Assuming the rule description is correct, operator overloading should not apply.  Returning `this` from an operator overload does not devalue the usage of the cascade operator.

**Comment by bwilkerson** [comment]
That's a much more compelling argument, thanks.
